javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala

```diff
@@ -0,0 +1,288 @@
+      } else { // e < 0
+        smallPowerOfTens(r + TableSize).mul(largePowerOfTens((q - 1) + LargeTableRange))
+      }
+    }
+  }
+
+  private def algorithmM(f: BigInteger, e: Int,
+      fmt: FloatingPointFormat): fmt.Repr = {
+    val MinSig = BigInteger.ZERO.setBit(fmt.SigBits - 1)
+    val MaxSig = BigInteger.ZERO.setBit(fmt.SigBits)
+
+    val MaxExp = (1 << fmt.ExpBits - 1) - 1 - fmt.ExplicitSigBits
+    val MinExp = -(1 << fmt.ExpBits - 1) + 2 - fmt.ExplicitSigBits // min exp for normal
+
+    // Approximate decimal f*10^e by binary (u/v)*2^k
+    // starting from k=0, iterate until u/v is normalized.
+    @tailrec def loop(u: BigInteger, v: BigInteger, k: Int): fmt.Repr = {
+      val divRem = u.divideAndRemainder(v)
+      val q = divRem(0)
+      val rem = divRem(1)
+      if (k == MinExp) {
+        if (q.compareTo(MinSig) >= 0 && q.compareTo(MaxSig) < 0) {
+          val z = fmt.toIEEE754(FloatingPoint(q, k))
+          roundByRemainder(q, rem, v, z)
+        } else {
+          underflow(q, v, rem)
+        }
+      } else if (k > MaxExp) {
+        fmt.PositiveInfinity
+      } else if (q.compareTo(MinSig) >= 0 && q.compareTo(MaxSig) < 0) {
+        val z = fmt.toIEEE754(FloatingPoint(q, k))
+        roundByRemainder(q, rem, v, z)
+      } else if (q.compareTo(MinSig) < 0) {
+        loop(u.multiply(BigInteger.TWO), v, k - 1)
+      } else { // x > 2^m
+        loop(u, v.multiply(BigInteger.TWO), k + 1)
+      }
+    }
+
+    // u = v*q + r (0 <= r < v)
+    // u/v = q + r/v (0 <= r/v < 1)
+    // therefore, q <= u/v < q+1
+    // if u/v < q + 1/2 then round down
+    // if u/v > q + 1/2 then round up
+    //
+    // u/v < q + 1/2 <=> q + r/v < q + 1/2 <=>
+    // r/v < 1/2 <=> r < r/2 <=> 2r < v <=> r < v - r
+    def roundByRemainder(q: BigInteger, rem: BigInteger, v: BigInteger,
+        z: fmt.Repr): fmt.Repr = {
+      val vMinusR = v.subtract(rem)
+      if (rem.compareTo(vMinusR) < 0) z
+      else if (rem.compareTo(vMinusR) > 0) fmt.nextFloat(z)
+      else if (q.remainder(BigInteger.TWO).intValue() == 0) z // even
+      else fmt.nextFloat(z)
+    }
+
+    @inline
+    def underflow(q: BigInteger, v: BigInteger, rem: BigInteger): fmt.Repr = {
+      if (q.compareTo(MinSig) < 0) { // mantissa is small enough
+        roundByRemainder(q, rem, v, fmt.reinterpretBits(q))
+      } else {
+        throw new Error("Shouldn't reach here.")
+      }
+    }
+
+    if (e < 0) loop(f, BigInteger.TEN.pow(-e), 0)
+    else loop(f.multiply(BigInteger.TEN.pow(e)), BigInteger.ONE, 0)
+  }
+
+  private def algorithmR(f: BigInteger, e: Int, z0: FloatingPoint,
+      fmt: FloatingPointFormat): fmt.Repr = {
+    @tailrec
+    def loop(z: fmt.Repr): fmt.Repr = {
+      val (m, exp) = fmt.frexp(z)
+      val mantissa = BigInteger.valueOf(m)
+      val (x, y) = {
+        if (e >= 0) {
+          if (exp >= 0)
+            (f.multiply(BigInteger.TEN.pow(e)), mantissa.multiply(BigInteger.ZERO.setBit(exp)))
```

Comment:
```suggestion
            (f.multiply(BigInteger.TEN.pow(e)), mantissa.shiftLeft(exp))
```
also in the other 3 below.